File: speller/sentimental-readability/readability.py
def get_average_letters_per_100_words(s):
    total_words = 0
    total_letters = 0
    for char in s:
        if char.isalpha():
            total_letters += 1
        elif char.isspace():
            total_words += 1

    total_words += 1
    average_letters_per_word = total_letters / total_words
    average_letters_per_100_words = average_letters_per_word * 100
    return average_letters_per_100_words

def get_average_sentences_per_100_words(s):
    num_sentences = 0
    total_words = 0
    for char in s:
        if char in ['.', '?', '!']:
            num_sentences += 1
        elif char.isspace():
            total_words += 1

    total_words += 1
    average_sentences_per_100_words = (num_sentences / total_words) * 100
    return average_sentences_per_100_words

def calculate_readability_score(s):
    letters_per_100_words = get_average_letters_per_100_words(s)
    sentences_per_100_words = get_average_sentences_per_100_words(s)

    score = 0.0588 * letters_per_100_words - 0.296 * sentences_per_100_words - 15.8

    if score - int(score) >= 0.5:
        return int(score) + 1
    else:
        return int(score)

input_text = input("Enter the text: ")
score = calculate_readability_score(input_text)
if score < 1:
    print("Before Grade 1")
elif score >= 16:
    print("Grade 16+")
else:
    print(f"Grade {score}")

File: readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

float getAverageLettersPer100Words(string s);
float getAverageSentencesPer100Words(string s);
int calculateReadabilityScore(string s);

int main()
{
    string input = get_string("Enter the text: ");
    int score = calculateReadabilityScore(input);
    if (score < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (score >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", score);
    }
}

float getAverageLettersPer100Words(string s)
{
    int totalWords = 0;
    int totalLetters = 0;

    for (int i = 0, length = strlen(s); i < length; i++)
    {
        if (isalpha(s[i]))
        {
            totalLetters++;
        }
        else if (isspace(s[i]))
        {
            totalWords++;
        }
    }

    // Increment total words by 1 for the last word
    totalWords++;

    // Calculate average letters per word
    float averageLettersPerWord = (float) totalLetters / totalWords;

    // Calculate average letters per 100 words
    float averageLettersPer100Words = averageLettersPerWord * 100;

    return averageLettersPer100Words;
}

float getAverageSentencesPer100Words(string s)
{
    int noOfSentences = 0;
    int totalWords = 0;

    for (int i = 0, length = strlen(s); i < length; i++)
    {
        if (s[i] == '.' || s[i] == '?' || s[i] == '!')
        {
            noOfSentences++;
        }
        else if (isspace(s[i]))
        {
            totalWords++;
        }
    }

    // Increment total words by 1 for the last word
    totalWords++;

    float averageSentencesPer100Words = ((float) noOfSentences / totalWords) * 100;

    return averageSentencesPer100Words;
}

int calculateReadabilityScore(string s)
{

    float lettersPer100Words = getAverageLettersPer100Words(s);
    float sentencesPer100Words = getAverageSentencesPer100Words(s);

    float score = 0.0588 * lettersPer100Words - 0.296 * sentencesPer100Words - 15.8;

    float decimalPart = score - (int) score;
    if (decimalPart >= 0.5)
    {
        return ceil(score);
    }
    else
    {
        return floor(score);
    }
}
